File: WolfDen/wwwroot/ts/Piece.ts
namespace TurnBasedBoardGame {
    export class Piece implements IExecuteMove {

        Node: Node;
        public Behaviours: any = {};
        Id: string;
        Owner: AdvanceWarsPlayer;
        Game: Game;

        get BehaviourList(): {}[] {
            return Object.values(this.Behaviours);;
        }

        constructor() {

        }

        ExecuteMove(move: any): boolean {
            var behaviour = this.Behaviours[move.id] as IExecuteMove;
            return behaviour.ExecuteMove(move.data);
        }

        Destroy() {
            delete this.Game.Pieces[this.Id];
            this.Owner.Pieces.splice(this.Owner.Pieces.indexOf(this), 1);
            this.Node.Piece = null;
        }
    }
}

File: WolfDen/Game/Piece.cs
using System;
using System.Collections.Generic;
using WolfDen.Game.Interfaces;

namespace WolfDen.Game
{
    public class Piece : IExecuteMove
    {
        public Node Node { get; set; }
        public Dictionary<string, IBehaviour> Behaviours = new Dictionary<string, IBehaviour>();
        public string Id;
        public Game Game { get; set; }
        public AdvanceWarsPlayer Owner { get; set; }

        public Piece(Game game)
        {
            Game = game;
        }

        public virtual bool ExecuteMove(dynamic data)
        {
            try
            {
                string id = data.id;
                return Behaviours[id].ExecuteMove(data.data);
            }
            catch (Exception ex)
            {
                return false;
            }
        }

        public void Destroy()
        {
            Game.Pieces.Remove(Id);
            Owner?.Pieces?.Remove(this);
            Node.Piece = null;
            foreach (var behaviour in Behaviours.Values)
            {
                behaviour.Destroy();
            }
        }
    }
}
